### PR TITLE
[MonologBridge] Add TokenProcessor

### PR DESCRIPTION
**Pull request by maidmaid**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/8156

``TokenProcessor`` is an out-of-the-box class that adds token in extra info log, like this:

```
[2016-12-29 14:48:09] app.INFO: admin connected! [] {"token":{
    "username":"admin",
    "authenticated":true,
    "roles":["ROLE_ADMIN"]}
}
```

### CODE REVIEW

**Comment by maidmaid** [comment]
If this feature will be merged, is it a good idea to register this processor in ``MonologBundle`` for all logs? Or else, maybe simply write a doc that explain its registration (containing something like [this](https://github.com/maidmaid/monolog-bundle/commit/b8341bdec47dfea0f590d9b2b87bd1d8369da409))?

**Comment by iltar** [comment]
Would this add the token to every log line? Because I don't want to see the token added 10x if my log contains 10 log lines to be honest.

The idea is great though, I'm using a custom formatter for the logging at the moment and I rather have this in the core.

**Comment by lyrixx** [comment]
@iltar Yes it will. That's how processor works.

But:

* This processor is not enabled by default
* You can add process on specific channel and/or handler.

**Comment by maidmaid** [comment]
I have improved token normalization that now is more json-like/readable:

```
[2016-12-29 14:48:09] app.INFO: admin connected! [] {"token":{
    "username":"admin",
    "auth.":true,
    "roles":["ROLE_ADMIN"]}
}
```

File: src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php [hunk 1]
```diff
@@ -29,8 +30,19 @@ public function __construct(TokenStorage $tokenStorage)
...
+        if ('cli' === php_sapi_name()) {
+            return $records;
+        }
+
+        $records['extra']['token'] = null;
         if (null !== $token = $this->tokenStorage->getToken()) {
-            $records['extra']['token'] = (string) $token;
+            $records['extra']['token'] = array(
+                'username' => $token->getUsername(),
+                'auth.' => $token->isAuthenticated(),
```
Comment (iltar) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: I suggest to call this key `authenticated`

Comment (fabpot) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: Indeed

File: src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php [hunk 2]
```diff
@@ -25,6 +25,7 @@ public function __invoke(array $record)
             'timestamp' => $record['datetime']->getTimestamp(),
             'message' => $record['message'],
             'priority' => $record['level'],
+
```
Comment (fabpot) [inline src/Symfony/Bridge/Monolog/Processor/DebugProcessor.php]: should be reverted

File: src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        $records['extra']['token'] = null;
+        if (null !== $token = $this->tokenStorage->getToken()) {
+            $records['extra']['token'] = array(
+                'username' => $token->getUsername(),
+                'auth.' => $token->isAuthenticated(),
+                'roles' => array_map(function (Role $role) {
+                    return $role->getRole();
+                }, $token->getRoles()),
```
Comment (fabpot) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: should be on one line

**Comment by maidmaid** [comment]
Thank you for your review @fabpot and @iltar. Changes are done.

File: src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bridge\Monolog\Processor;
+
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorage;
+use Symfony\Component\Security\Core\Role\Role;
+
+/**
+ * TokenProcessor adds token extra info.
```
Comment (xabbuh) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: "Adds the current security token to the log entry."?

Comment (maidmaid) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: Indeed, it's better.

File: src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                'username' => $token->getUsername(),
+                'authenticated' => $token->isAuthenticated(),
+                'roles' => array_map(function (Role $role) { return $role->getRole(); }, $token->getRoles()),
+            );
+        }
+
+        return $records;
+    }
+
+    public function isDeniedProcess()
```
Comment (xabbuh) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: Is this method really needed?

Comment (maidmaid) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: In CLI mode, ``extra.token`` should not be added, for example in commands but not in tests. This method allows [to mock this behavior](https://github.com/symfony/symfony/pull/21086/commits/653c24455d8689671e450f4c84b1d2bf017ab6b8#diff-e846e2de8865a11af799406eac1d2bdeR37) in tests.

Comment (ro0NL) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: > In CLI mode, extra.token should not be added

Why not?

Comment (xabbuh) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: I was more referring to this being a dedicated method while it could easily be inlined.

Comment (xabbuh) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: Furthermore, a method being `public` is an extension point which we would have to maintain.

Comment (maidmaid) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: Indeed, this is not very helpful. I have removed this method.

**Comment by xabbuh** [comment]
You need to add the `security/core` package as a dependency to the `require-dev` section in the `composer.json` file of MonologBridge.

File: src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public function __invoke(array $records)
+    {
+        $records['extra']['token'] = null;
+        if (null !== $token = $this->tokenStorage->getToken()) {
+            $records['extra']['token'] = array(
+                'username' => $token->getUsername(),
+                'authenticated' => $token->isAuthenticated(),
+                'roles' => array_map(function (Role $role) { return $role->getRole(); }, $token->getRoles()),
```
Comment (ro0NL) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: Should be `RoleInterface`

Comment (maidmaid) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: ``RoleInterface`` is now deprecated

**Comment by maidmaid** [comment]
I updated the ``composer.json`` @xabbuh.

File: src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php (hunk 6 above)
Comment (ro0NL) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: And not yet removed, they are still returned from `getRoles`.

Comment (maidmaid) [inline src/Symfony/Bridge/Monolog/Processor/TokenProcessor.php]: :+1: it's done

**Comment by maidmaid** [comment]
I think that failed tests are not related.

**Comment by maidmaid** [comment]
Review is ok?

**Comment by maidmaid** [comment]
Need some help to fix tests please.

**Comment by maidmaid** [comment]
AppVeyor break tests are related?

**Comment by sstok** [comment]
AppVeyor Failure seems unrelated.

**Comment by maidmaid** [comment]
I have repushed last commit, and tests are now green :)

**Comment by maidmaid** [comment]
PR ready.

**Comment by maidmaid** [comment]
PR ready to merge since some time :)

**Review by Simperfit — APPROVED** [review]

**Comment by lyrixx** [comment]
:+1:

File: src/Symfony/Bridge/Monolog/composer.json [hunk 7]
```diff
@@ -23,6 +23,7 @@
     "require-dev": {
         "symfony/console": "~2.8|~3.0|~4.0",
         "symfony/event-dispatcher": "~2.8|~3.0|~4.0",
+        "symfony/security-core": "~2.8|~3.0",
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/composer.json]: missing `|~4.0`

Comment (maidmaid) [inline src/Symfony/Bridge/Monolog/composer.json]: ``RoleInterface`` is used here, and will be removed in 4.0.

Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/composer.json]: then I suggest to remove the type int, it's purely internal and does justify a deps limitation

Comment (maidmaid) [inline src/Symfony/Bridge/Monolog/composer.json]: Type hint removed.

**Comment by nicolas-grekas** [comment]
Thank you @maidmaid.

**Pull request merged**